astropy/config/paths.py

```diff
@@ -19,6 +20,47 @@
 ]
 
 
+class _DirectoryType(Enum):
+    CONFIG = auto()
+    CACHE = auto()
+
+
+def _get_dir_path(rootname: str, dirtype: _DirectoryType) -> Path:
+    # symlink will be set to this if the directory is created
+    linkto = None
+
+    match dirtype:
+        case _DirectoryType.CONFIG:
+            cls = set_temp_config
+            fallback = "config"
+        case _DirectoryType.CACHE:
+            cls = set_temp_cache
+            fallback = "cache"
+        case _:
+            raise ValueError(f"Unexpected {dirtype=}")
+
+    # If using set_temp_x, that overrides all
+    if cls._temp_path is not None:
+        xch = cls._temp_path
+        path = xch / rootname
+        if not path.is_file():
```

Comment:
note that this condition didn't exist in `get_config_dir`, but this implementation seems to pass all tests (while unconditionally calling `path.mkdir` doesn't), so I figured it's preferable to special casing cache.